docs/features/source-generators.cookbook.md

```diff
@@ -0,0 +1,250 @@
+# Source Generators Cookbook
+
+## Summary
+
+This document aims to be a guide to help the creation of source generators by providing a series of reusable patterns.
+It also aims to set out what types of generators are possible under the current design and what is explictly out of scope.
+
+> **Note**: The design for the source generator proposal is still under review. This document uses only one possible syntax, and
+> it is expected to change without notice as the feature evolves.
+
+## Proposal
+
+It is worth breifly setting out the design goals of generators here. For a more complete design see the [full design document](source-generators.md)
+
+- Generators produce one or more strings that represent C# source code to be added to the compilation.
+- Explicitly _additive_ only. Generators can add new code to a compilation but may **not** modify existing user code.
+- May access _additional files_, that is, non-C# source texts.
+- Run _un-ordered_, each generator will see the same input compilation, with no access to files created by other source generators.
+- A user specifies the generators to run via list of assemblies, much like analyzers.
+
+## Out of scope designs
+
+We will briefly look at the non-solvable problems as examples of the kind of problems source generators are *not* designed to solve:
+
+### Language features
+
+Source generators are not designed to replace new language features: for instance one could imagine [records](records.md) being implemented as a source generator
+that converts the specified syntax to a compilable C# representation.
+
+We explicitly consider this to be an anti-pattern; the language will continue to evolve and add new features, and we don't expect source generators to be
+a way to enable this. Doing so would create new 'dialects' of C# that are incompatible with the compiler without generators. Further, because generators,
+by design, cannot interact with each other, language features implemented in this way would quickly become incompatible with other additions to the language.
+
+### Code-not-source generators
```

Comment:
Where did we come up with "code-not-source" terms? In my head i'd always thought of this as mutating generators. Generators that either mutate IL or C# code. 

I'm not against this term but digesting it at the moment.

Reply:
I like the phrasing that @chsienki is using below, which is:

>  They are, by defintion, code altering operations which is explicitly ruled out by the source generation proposal.

So maybe change this to "Code alterations" or "Code rewriters"?

To me, code gen sounds as much additive as source gen.